From a31ba48d0e3f1af93d4afc8db46cce4265526786 Mon Sep 17 00:00:00 2001
From: mrxx0 <0mrxx0@proton.me>
Date: Sat, 15 Mar 2025 14:30:23 +0100
Subject: [PATCH] NFC auto turn off

Ported to AXP.OS lineage-22.2 (rhodep fork) from GrapheneOS ebce750e7455.

Change-Id: Ibd74cff5ef2f849b34d295ac251685ff025b5682
---
 core/java/android/provider/Settings.java      |  6 ++
 .../com/android/server/ext/NfcAutoOff.java    | 73 +++++++++++++++++++
 .../android/server/ext/SystemServerExt.java   |  4 +
 3 files changed, 83 insertions(+)
 create mode 100644 services/core/java/com/android/server/ext/NfcAutoOff.java

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index b221e1d44..2b196face 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -19609,6 +19609,12 @@ public final class Settings {
          */
         public static final String BLUETOOTH_OFF_TIMEOUT = "bluetooth_off_timeout";
 
+        /**
+         * The amount of time in milliseconds before an enabled NFC adapter is turned off
+         * @hide
+         */
+        public static final String NFC_OFF_TIMEOUT = "nfc_off_timeout";
+
         /**
          * Whether repair mode is active on the device.
          * <p>
diff --git a/services/core/java/com/android/server/ext/NfcAutoOff.java b/services/core/java/com/android/server/ext/NfcAutoOff.java
new file mode 100644
index 000000000..9593e3415
--- /dev/null
+++ b/services/core/java/com/android/server/ext/NfcAutoOff.java
@@ -0,0 +1,73 @@
+package com.android.server.ext;
+
+import android.annotation.Nullable;
+import android.nfc.NfcAdapter;
+import android.provider.Settings;
+import android.content.pm.PackageManager;
+import android.content.IntentFilter;
+import android.content.Intent;
+import android.content.Context;
+import android.content.BroadcastReceiver;
+
+import android.util.Slog;
+
+class NfcAutoOff extends DelayedConditionalAction {
+    private static final String TAG = NfcAutoOff.class.getSimpleName();
+
+    private final Context context;
+
+    NfcAutoOff(SystemServerExt sse) {
+        super(sse, sse.bgHandler);
+        context = sse.context;
+    }
+
+    @Override
+    protected boolean shouldScheduleAlarm() {
+        return isAdapterOn();
+    }
+
+    @Override
+    protected void alarmTriggered() {
+        if (isAdapterOn()) {
+            Slog.d(TAG, "adapter.disable(true)");
+            getAdapter().disable(true);
+        }
+    }
+
+    @Override
+    protected void registerStateListener() {
+        IntentFilter f = new IntentFilter();
+        f.addAction(NfcAdapter.ACTION_ADAPTER_STATE_CHANGED);
+
+        sse.context.registerReceiverForAllUsers(new BroadcastReceiver() {
+            @Override
+            public void onReceive(Context broadcastContext, Intent intent) {
+                Slog.d(TAG, "" + intent + ", extras " + intent.getExtras().deepCopy());
+                update();
+            }
+        }, f, null, handler);
+    }
+
+    @Nullable
+    private NfcAdapter getAdapter() {
+        return NfcAdapter.getDefaultAdapter(context);
+    }
+
+    private boolean isAdapterOn() {
+        Slog.d(TAG, "isAdapterOn");
+        var adapter = getAdapter();
+        if (adapter != null) {
+            boolean isEnabled = adapter.isEnabled();
+            Slog.d(TAG, "isEnabled: " + isEnabled);
+            return isEnabled;
+        } else {
+            Slog.d(TAG, "adapter is null");
+        }
+        return false;
+    }
+
+    @Override
+    protected String getDelayGlobalSettingsKey() {
+        return Settings.Global.NFC_OFF_TIMEOUT;
+    }
+}
diff --git a/services/core/java/com/android/server/ext/SystemServerExt.java b/services/core/java/com/android/server/ext/SystemServerExt.java
index e543b622a..264c67dc2 100644
--- a/services/core/java/com/android/server/ext/SystemServerExt.java
+++ b/services/core/java/com/android/server/ext/SystemServerExt.java
@@ -59,6 +59,10 @@ public final class SystemServerExt {
         if (packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH, 0)) {
             new BluetoothAutoOff(this);
         }
+
+        if (packageManager.hasSystemFeature(PackageManager.FEATURE_NFC, 0)) {
+            new NfcAutoOff(this);
+        }
     }
 
     public void registerReceiver(BroadcastReceiver receiver, IntentFilter filter, Handler handler) {
-- 
2.43.0

